### PR TITLE
Apollo list transactions

### LINKED ISSUE #822: 🔧 [Refactor] Implement Community Call listTransactions on Apollo
## 🔧 Refactor ticket

Implement Community Call listTransactions on Apollo

### PR DESCRIPTION
**Pull request by einhornimmond**
## 🍰 Pullrequest
Implement listTransactions call from Community Server into Apollo

### Issues
- fixes #822 

### Todo 
- [x] load gdt sum and gdd balance with decay
- [x] add UserTransaction Model for table state_user_transactions
- [x] implement paging 
- [x] add Transactions Model for table transactions
- [x] add TransactionSendCoin Model for table transaction_send_coins
- [x] add TransactionCreation Model for table transaction_creations
- [x] implement calculating decay transactions and add them

### CODE REVIEW

**Comment by ulfgebhardt** [comment]
This looks very good. What is missing?

**Comment by einhornimmond** [comment]
Decay don't work properly

File: backend/src/graphql/resolvers/listTransactions.ts [hunk 1]
```diff
@@ -0,0 +1,194 @@
...
+    // balance
+    if (userTransaction.transactionTypeId === 1) {
+      // creation
+      const creation = transaction.transactionCreation
+
+      finalTransaction.name = 'Gradido Akademie'
+      finalTransaction.type = 'creation'
+      // finalTransaction.targetDate = creation.targetDate
+      finalTransaction.balance = roundFloorFrom4(creation.amount)
+    } else if (userTransaction.transactionTypeId === 2) {
```
Comment (ulfgebhardt) [inline backend/src/graphql/resolvers/listTransactions.ts:101]:
```suggestion
    } else if (userTransaction.transactionTypeId === TRANSACTION_TYPES.TYPE_2) {

File: backend/src/graphql/resolvers/listTransactions.ts [hunk 2]
```diff
@@ -0,0 +1,194 @@
...
+      // send coin
+      const sendCoin = transaction.transactionSendCoin
+      let otherUser: dbUser | undefined
+      finalTransaction.balance = roundFloorFrom4(sendCoin.amount)
+      if (sendCoin.userId === user.id) {
+        finalTransaction.type = 'send'
```
Comment (ulfgebhardt) [inline backend/src/graphql/resolvers/listTransactions.ts:107]:
```suggestion
        finalTransaction.type = TRANSACTION_TYPE_NAMES.SEND
```

File: backend/src/graphql/resolvers/listTransactions.ts [hunk 3]
```diff
@@ -0,0 +1,194 @@
...
+        otherUser = userIndiced[sendCoin.recipiantUserId]
+        // finalTransaction.pubkey = sendCoin.recipiantPublic
+      } else if (sendCoin.recipiantUserId === user.id) {
+        finalTransaction.type = 'receive'
```
Comment (ulfgebhardt) [inline backend/src/graphql/resolvers/listTransactions.ts:111]: enum pls

File: backend/src/graphql/resolvers/listTransactions.ts [hunk 4]
```diff
@@ -0,0 +1,194 @@
...
+    }
+  }
+
+  return finalTransactions
+}
+
+export default async function listTransactions(
+  firstPage: number,
+  items: number,
+  order: 'ASC' | 'DESC',
```
Comment (ulfgebhardt) [inline backend/src/graphql/resolvers/listTransactions.ts:152]: enum pls compatible with the graphql

File: backend/src/graphql/resolvers/listTransactions.ts [hunk 5]
```diff
@@ -0,0 +1,194 @@
...
+): Promise<TransactionList> {
+  let limit = items
+  let offset = 0
+  let skipFirstTransaction = false
+  if (firstPage > 1) {
+    offset = (firstPage - 1) * items - 1
+    limit++
+  }
+
+  if (offset && order === 'ASC') {
```
Comment (ulfgebhardt) [inline backend/src/graphql/resolvers/listTransactions.ts:163]: enum

File: backend/src/graphql/resolvers/listTransactions.ts [hunk 6]
```diff
@@ -0,0 +1,194 @@
...
+    user.id,
+    limit,
+    offset,
+    order,
+  )
+  skipFirstTransaction = userTransactionsCount > offset + limit
+  const decay = !(firstPage > 1)
+  let transactions: Transaction[] = []
+  if (userTransactions.length) {
+    if (order === 'DESC') {
```
Comment (ulfgebhardt) [inline backend/src/graphql/resolvers/listTransactions.ts:176]: enum

File: backend/src/graphql/resolvers/listTransactions.ts [hunk 7]
```diff
@@ -0,0 +1,194 @@
...
+      userTransactions = userTransactions.reverse()
+    }
+    transactions = await calculateAndAddDecayTransactions(
+      userTransactions,
+      user,
+      decay,
+      skipFirstTransaction,
+    )
+    if (order === 'DESC') {
```
Comment (ulfgebhardt) [inline backend/src/graphql/resolvers/listTransactions.ts:185]: enum

File: backend/src/typeorm/entity/UserTransaction.ts [hunk 8]
```diff
@@ -0,0 +1,36 @@
...
+  balance: number
+
+  @Column({ name: 'balance_date', type: 'timestamp' })
+  balanceDate: Date
+
+  static findByUserPaged(
+    userId: number,
+    limit: number,
+    offset: number,
+    order: 'ASC' | 'DESC',
```
Comment (ulfgebhardt) [inline backend/src/typeorm/entity/UserTransaction.ts:27]: enum

File: backend/src/util/decay.ts [hunk 9]
```diff
@@ -1,4 +1,55 @@
...
+}
+
+async function calculateDecayWithInterval(
+  amount: number,
+  from: number | Date,
+  to: number | Date,
+): Promise<Decay> {
+  const decayStartBlock = await Decay.getDecayStartBlock()
+
+  const result = new Decay(undefined)
```
Comment (ulfgebhardt) [inline backend/src/util/decay.ts:25]: can we use `new Decay()`instead of explicit `undefined` pls

File: backend/src/util/decay.ts [hunk 10]
```diff
@@ -1,4 +1,55 @@
...
+
   const decayDuration = (to.getTime() - from.getTime()) / 1000
-  return amount * Math.pow(0.99999997802044727, decayDuration)
+  return decayFormula(amount, decayDuration)
+}
+
+async function calculateDecayWithInterval(
+  amount: number,
+  from: number | Date,
+  to: number | Date,
```
Comment (ulfgebhardt) [inline backend/src/util/decay.ts:21]: require date only?

File: backend/src/util/decay.ts [hunk 11]
```diff
@@ -1,4 +1,55 @@
...
+  result.balance = amount
+  const fromMillis = typeof from === 'number' ? from : from.getTime()
+  const toMillis = typeof to === 'number' ? to : to.getTime()
+  result.decayStart = (fromMillis / 1000).toString()
+  result.decayEnd = (toMillis / 1000).toString()
+
+  // (amount, from.getTime(), to.getTime())
+
+  // if no decay start block exist or decay startet after end date
+  if (decayStartBlock === undefined || decayStartBlock.received.getTime() > toMillis) {
```
Comment (ulfgebhardt) [inline backend/src/util/decay.ts:35]:
```suggestion
  if (!decayStartBlock || decayStartBlock.received.getTime() > toMillis) {
```

**Review by ulfgebhardt — CHANGES_REQUESTED** [review]

**Review by ulfgebhardt — APPROVED** [review]

**Pull request merged**